DSharpPlus/Utilities.cs

```diff
@@ -169,12 +169,154 @@ internal static IEnumerable<ulong> GetChannelMentions(DiscordMessage message)
+                            var role = message.Channel.Guild.GetRole(tagId);
+                            return await objectReplacer(role);
+                        }
+                        return await objectReplacer(null);
+
+                    case "#": // Channel mention
+                        if (message.Channel?.Guild != null)
+                        {
+                            var channel = message.Channel.Guild.GetChannel(tagId);
+                            return await objectReplacer(channel);
+                        }
+                        return await objectReplacer(null);
+
+                    case string s when s.StartsWith("a:") || s.StartsWith(":"): // Guild emoji
+                        var exists = DiscordEmoji.TryFromGuildEmote(message.Discord, tagId, out var emoji);
+                        // We still have enough information to construct a dummy DiscordEmoji
+                        // which could provide us for example an Url for the emoji
+                        // This could happen if eg a Nitro user used an emoji from a server
+                        // where the bot is not present
+                        // We explicitely mark as IsAvailable = false to avoid it's usage
+                        emoji = exists ? emoji : new DiscordEmoji
+                        {
+                            Id = tagId,
+                            Name = m.Groups["TagName"].Value,
+                            IsAvailable = false,
+                            IsAnimated = s.StartsWith("a:"),
+                            Discord = message.Discord,
+                            RequiresColons = true
+                        };
+                        return await objectReplacer(emoji);
+
+                    default:
+                        throw new ArgumentException("Matched something unexpected");
+                }
+            });
+
+            // Now we take care of common emojis
+            // We sort them so we first replace the most complex ("composite") ones first
+
+            // Converting the ":emojiname:" format
+            var emojis = DiscordEmoji.UnicodeEmojis.Keys
+                .Where(e => text.Contains(e))
+                .OrderByDescending(e => e.Count(c => c == ':'))
+                .ThenByDescending(e => e.Length)
+                .Select(e => new { Token = e, Emoji = DiscordEmoji.FromName(message.Discord, e, false) })
+                .Union(// Converting the unicode emojis
+                    DiscordEmoji.DiscordNameLookup.Keys
+                    .Where(e => text.Contains(e))
+                    .OrderByDescending(e => e.Length)
+                    .Select(e => new { Token = e, Emoji = DiscordEmoji.FromUnicode(e) }));
+
+            foreach (var emojiInfo in emojis)
+            {
+                var replacement = await objectReplacer(emojiInfo.Emoji);
+                text = text.Replace(emojiInfo.Token, replacement);
+            }
+
+            return text;
```

Comment:
This is an abomination. Simply escaping both colons and stripping the Id (if applicable) would be infinitely faster and more readable.

Reply:
The ID here is not applicable, the Ided guild emotes were parsed already at this point, this is for common ones.
I wanted to make sure I catch all the emojis so they can be all passed down to the callback for user-defined replacement, and not catching something which shouldn't (eg :somerandomtext: not being an emote).

Reply:
If you use Regex.Replace it doesn't matter if it's a valid emote or not because the point isn't to check if an emote is valid, if you're just trying to humanize what could be an emote

Reply:
It's a little complex than that. I need to make sure that the output is as close as possible to what is shown by Discord itself to humans. So if I replace an invalid emote with something which isn't replaced by Discord, it wouldn't be correct.

Reply:
Discord renders the emoji, so. And escaping an invalid emoji produces the same result as escaping a valid emoji.

Reply:
Well, I ditched all that code. ¯\\\_(ツ)\_/¯